In [121]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix,accuracy_score
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import cross_val_score, cross_val_predict

In [75]:
# Wczytanie zbioru danych oraz nazwanie kolumn zgodnie z opisem ( i., ii.)
wines = pd.read_table("wine.data",sep=',',names=["Class","Alcochol","Malic acid","Ash","Alcalinity of ash","Magnesium","Total phenols","Flavaniods","Nonflavanoid phenols","Proanthocyanins"
             ,"Color intensity","Hue","OD280/OD315 of diluted wines","Proline"])

In [76]:
wines.head()

Class  Alcochol  Malic acid   Ash  Alcalinity of ash  Magnesium  \
0      1     14.23        1.71  2.43               15.6        127   
1      1     13.20        1.78  2.14               11.2        100   
2      1     13.16        2.36  2.67               18.6        101   
3      1     14.37        1.95  2.50               16.8        113   
4      1     13.24        2.59  2.87               21.0        118   

   Total phenols  Flavaniods  Nonflavanoid phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

In [99]:
# Ograniczenie zbioru do odpowiednio: wszystkich składowych/ 2 pierwszych/ 5 pierwszych/ 10 pierwszych (iii., iv.)
X = [("Pełny zbiór danych:",wines.iloc[:,1:].values),("Ograniczenie do 2 pierwszych składowych:",wines.iloc[:,1:3].values)
            ,("Ograniczenie do 5 pierwszych składowych:",wines.iloc[:,1:6].values)
            ,("Ograniczenie do 10 pierwszych składowych:",wines.iloc[:,1:11].values)]
y = wines.iloc[:,0].values

#### Narazie zakładmy, że cały zbiór to PU i PT jednocześcnie (powtórne podstawienie), dlatego możemy przypusić, że taka klasyfikacja jest dosyć optymistyczna. W następnej części zadnia dokonamy podziału całego zbioru na PU i PT w odpowienich stosunkach

In [156]:
#LDA
for name,x in X:
    lda = LinearDiscriminantAnalysis()
    lda.fit(x,y)
    pred_lda = lda.predict(x)
    print(name)
    print("Accuracy: ",accuracy_score(y, pred_lda))
    print("Macierz pomyłek: \n",confusion_matrix(y, pred_lda),"\n")

Pełny zbiór danych:
Accuracy:  1.0
Macierz pomyłek: 
 [[59  0  0]
 [ 0 71  0]
 [ 0  0 48]] 

Ograniczenie do 2 pierwszych składowych:
Accuracy:  0.8089887640449438
Macierz pomyłek: 
 [[51  1  7]
 [ 5 61  5]
 [ 7  9 32]] 

Ograniczenie do 5 pierwszych składowych:
Accuracy:  0.8764044943820225
Macierz pomyłek: 
 [[54  1  4]
 [ 1 65  5]
 [ 3  8 37]] 

Ograniczenie do 10 pierwszych składowych:
Accuracy:  0.9887640449438202
Macierz pomyłek: 
 [[59  0  0]
 [ 1 70  0]
 [ 0  1 47]] 



In [157]:
#QDA
for name,x in X:
    qda = QuadraticDiscriminantAnalysis()
    qda.fit(x,y)
    pred_qda = qda.predict(x)
    print(name)
    print("Accuracy: ",accuracy_score(y, pred_qda))
    print("Macierz pomyłek: \n",confusion_matrix(y, pred_qda),"\n")

Pełny zbiór danych:
Accuracy:  0.9943820224719101
Macierz pomyłek: 
 [[59  0  0]
 [ 1 70  0]
 [ 0  0 48]] 

Ograniczenie do 2 pierwszych składowych:
Accuracy:  0.8146067415730337
Macierz pomyłek: 
 [[52  1  6]
 [ 4 62  5]
 [ 7 10 31]] 

Ograniczenie do 5 pierwszych składowych:
Accuracy:  0.8876404494382022
Macierz pomyłek: 
 [[53  1  5]
 [ 1 65  5]
 [ 2  6 40]] 

Ograniczenie do 10 pierwszych składowych:
Accuracy:  0.9943820224719101
Macierz pomyłek: 
 [[59  0  0]
 [ 1 70  0]
 [ 0  0 48]] 



In [158]:
#NB
for name,x in X:
    nb = GaussianNB()
    nb.fit(x,y)
    pred_nb = nb.predict(x)
    print(name)
    print("Accuracy: ",accuracy_score(y, pred_nb))
    print("Macierz pomyłek: \n",confusion_matrix(y, pred_nb),"\n")

Pełny zbiór danych:
Accuracy:  0.9887640449438202
Macierz pomyłek: 
 [[58  1  0]
 [ 0 70  1]
 [ 0  0 48]] 

Ograniczenie do 2 pierwszych składowych:
Accuracy:  0.8089887640449438
Macierz pomyłek: 
 [[51  1  7]
 [ 4 62  5]
 [ 7 10 31]] 

Ograniczenie do 5 pierwszych składowych:
Accuracy:  0.8539325842696629
Macierz pomyłek: 
 [[52  3  4]
 [ 2 62  7]
 [ 3  7 38]] 

Ograniczenie do 10 pierwszych składowych:
Accuracy:  0.9606741573033708
Macierz pomyłek: 
 [[56  3  0]
 [ 2 68  1]
 [ 0  1 47]] 



#### Klasyfikatory osiagają lepszą skuteczność im więcej składowych jest uwzględnianych. Sposród LDA,QDA i NB nie ma zdecydowanego zwycięzcy, jednak to LDA osiągneło największą dokładnośc klasyfikacji. 

In [84]:
# Ograniczenie zbioru do 2 pierwszych składowych (v.)
X2 = wines.iloc[:,1:3].values
# Podzielenie zbioru na PU/PT w stosunku 75/25
X_train, X_test, y_train, y_test = train_test_split(X2, y, test_size=0.25, random_state=1)
# Podzielenie zbioru PU na PU/PW w stosunku 50/25
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.33, random_state=1)
# Tym samym otrzymaliśmy PU/PW/PT w stosunku 50/25/25

In [96]:
# Trenujmey wszystkie 3 klasyfikatory na PU a następnie sprawdzamy ich dokladność na PW

lda = LinearDiscriminantAnalysis()
lda.fit(X_train,y_train)
pred_lda = lda.predict(X_val)

qda = QuadraticDiscriminantAnalysis()
qda.fit(X_train,y_train)
pred_qda = qda.predict(X_val)

nb = GaussianNB()
nb.fit(X_train,y_train)
pred_nb = nb.predict(X_val)

print("\tAccuracy")
print("LDA: ",accuracy_score(y_val, pred_lda))
print("QDA: ",accuracy_score(y_val, pred_qda))
print("NB:  ",accuracy_score(y_val, pred_nb))

	Accuracy
LDA:  0.7727272727272727
QDA:  0.8409090909090909
NB:   0.8636363636363636


#### Na podstawie PW otrzymaliśmy, że klasyfikator NB jest najlepszy, dlatego do PT wybierzemy jego

In [128]:
nb = GaussianNB()
nb.fit(X_train,y_train)
pred_nb = nb.predict(X_test)
print("Accuracy: ",accuracy_score(y_test, pred_nb))

Accuracy:  0.8222222222222222


#### Skuteczność NB na PT spadła, dlatego możemy wywnioskować, że podział całego zbioru na PU/PW/PT w stosunku 50/25/25 nie jest optymalnym rozwiązaniem

In [150]:
# Ograniczenie zbioru do 2 pierwszych składowych (vi.)
lda = LinearDiscriminantAnalysis()
lda.fit(X2,y)

# Wykonujemy kroswalidacje na klasyfikatorze lda (k = 10)
y_pred = cross_val_predict(lda, X2, y, cv=10)
print("Accuracy: ",accuracy_score(y, y_pred))

Accuracy:  0.7640449438202247


#### Porównanie skuteczności:
#### 10-krotna kroswalidacja = 0.76
#### Powtórne podstawienie = 0.81
#### Poprzedni podpunkt = 0.78

#### Z otrzymanych wyników, możemy stwierdzić, że kroswalidacja wypadła najgorzej i nie podniosła skuteczności klasyfikaotra LDA 